libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,383 @@
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	"golang.org/x/crypto/pbkdf2"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/file"
+)
+
+const (
+	filePermission = 0600
+
+	// Encryption Related constants
+	iVLength        = 12
+	saltLength      = 64
+	iterationsCount = 10000
+	keyLength       = 32
+)
+
+// Version of the keystore format, will be added at the beginning of the file.
+var version = []byte("v1")
+
+// FileKeystore Allows to store key / secrets pair securely into an encrypted local file.
+type FileKeystore struct {
+	sync.RWMutex
+	Path     string
+	secrets  map[string]serializableSecureString
+	dirty    bool
+	password *SecureString
+}
+
+// Allow the original SecureString type to be correctly serialized to json.
+type serializableSecureString struct {
+	*SecureString
+	Value []byte `json:"value"`
+}
+
+// NewFileKeystore returns an new File based keystore or an error, currently users cannot set their
+// own password on the keystore, the default password will be an empty string. When the keystore
+// is initialied the secrets are automatically loaded into memory.
+func NewFileKeystore(keystoreFile string) (Keystore, error) {
+	return NewFileKeystoreWithPassword(keystoreFile, NewSecureString([]byte("")))
+}
+
+// NewFileKeystoreWithPassword return a new File based keystore or an error, allow to define what
+// password to use to create the keystore.
+func NewFileKeystoreWithPassword(keystoreFile string, password *SecureString) (Keystore, error) {
+	keystore := FileKeystore{
+		Path:     keystoreFile,
+		dirty:    false,
+		password: password,
+		secrets:  make(map[string]serializableSecureString),
+	}
+
+	err := keystore.load()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &keystore, nil
+}
+
+// Retrieve return a SecureString instance that will contains both the key and the secret.
+func (k *FileKeystore) Retrieve(key string) (*SecureString, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	secret, ok := k.secrets[key]
+	if !ok {
+		return nil, ErrKeyDoesntExists
+	}
+	return NewSecureString(secret.Value), nil
+}
+
+// Store add the key pair to the secret store and mark the store as dirty.
+func (k *FileKeystore) Store(key string, value []byte) error {
+	k.Lock()
+	defer k.Unlock()
+
+	k.secrets[key] = serializableSecureString{Value: value}
+	k.dirty = true
+	return nil
+}
+
+// Delete an existing key from the store and mark the store as dirty.
+func (k *FileKeystore) Delete(key string) error {
+	k.Lock()
+	defer k.Unlock()
+
+	delete(k.secrets, key)
+	k.dirty = true
+	return nil
+}
+
+// Save persists the in memory data to disk if needed.
+func (k *FileKeystore) Save() error {
+	k.Lock()
+	defer k.Unlock()
+
+	if k.dirty == false {
+		return nil
+	}
+
+	temporaryPath := fmt.Sprintf("%s.tmp", k.Path)
+	if _, err := os.Stat(temporaryPath); os.IsExist(err) {
```

Comment:
This isn't the correct way to check if a file exists (the logic is wrong, but that's not my point). The good thing is that you don't need this check. When you call `os.OpenFile` use `os.O_EXCL` instead of `os.O_TRUNC` and you'll get an error if the file exists (which you can test for with `os.IsExists()`).

This is better because because there no race condition with anything else that's writing the file between the time you call `os.Stat` and `os.OpenFile`.